Core: Bump C++ Standard to C++20

Repiteo opened the pull request:
- Supersedes #89660

In contrast to the above PR, which attempted to simultaneously support C++17 alongside later standards, this PR aims to make our baseline C++20 outright. The *immediate* benefit is removing the need for several workarounds via macros & `#if __cplusplus` wrappers, making for a much cleaner codebase overall. The much, much bigger benefit is the freedom to actually *integrate* C++20 features, allowing this PR to satisfy our contribution Best Practices & not exist as a solution looking for a problem.

**EDIT:** After a core meeting, it was instead decided that integrations of features would be best handled in entirely separate PRs, so that we can pick-and-choose which features are best to integrate. While that dampens the effect of this PR in isolation, it's now doing so with the knowledge that it will not exist in isolation and would soon be followed with proper integrations of the new features that the repo would directly benefit from. Future edits of this OP will point to specific discussions/implementations of said features up until this is merged.

AThousandShips commented:
Has the concerns with support for c++20 on niche compilers like those for various console ports been resolved? That has been the main thing holding this back other than just the workload

Repiteo commented:
I'm not certain if that's been resolved, and I certainly wouldn't want this merged before that was the case either. Things like Android will already need to be updated to accomodate C++20, so I know this wouldn't happen straight away no matter what. Regardless, this PR operating under the assumption that it *will* be supported; that way it can get a kick-start on actually integrating features & we won't have to do any `#ifdef` song-and-dance

AThousandShips commented:
Absolutely, generally absolutely in favor of adding various new QOL features from C++20, brought it up in the platforms channel and we'll see what information we can dig up

Mickeon commented on core/math/vector2.h:
```diff
@@ -243,12 +240,11 @@ _FORCE_INLINE_ Vector2 Vector2::operator-() const {
+std::partial_ordering Vector2::operator<=>(const Vector2 &p_other) const {
+	if (x <=> p_other.x == 0) {
+		return y <=> p_other.y;
+	}
+	return x <=> p_other.x;
```
I am a bit concerned on the new operator. I get it, but I am not sure it is an actual improvement readability-wise.

Repiteo replied:
It *does* take some time to get used to, I'll give you that. But the readability improvement comes more from consolidation of existing operators, which is more easily demonstrated on larger examples like Vector4:
<details>
<summary>Old</summary>

```cpp
bool Vector4::operator<(const Vector4 &p_v) const {
	if (x == p_v.x) {
		if (y == p_v.y) {
			if (z == p_v.z) {
				return w < p_v.w;
			}
			return z < p_v.z;
		}
		return y < p_v.y;
	}
	return x < p_v.x;
}

bool Vector4::operator>(const Vector4 &p_v) const {
	if (x == p_v.x) {
		if (y == p_v.y) {
			if (z == p_v.z) {
				return w > p_v.w;
			}
			return z > p_v.z;
		}
		return y > p_v.y;
	}
	return x > p_v.x;
}

bool Vector4::operator<=(const Vector4 &p_v) const {
	if (x == p_v.x) {
		if (y == p_v.y) {
			if (z == p_v.z) {
				return w <= p_v.w;
			}
			return z < p_v.z;
		}
		return y < p_v.y;
	}
	return x < p_v.x;
}

bool Vector4::operator>=(const Vector4 &p_v) const {
	if (x == p_v.x) {
		if (y == p_v.y) {
			if (z == p_v.z) {
				return w >= p_v.w;
			}
			return z > p_v.z;
		}
		return y > p_v.y;
	}
	return x > p_v.x;
}
```
</details>
<details>
<summary>New</summary>

```cpp
std::partial_ordering Vector4::operator<=>(const Vector4 &p_other) const {
	if (x <=> p_other.x == 0) {
		if (y <=> p_other.y == 0) {
			if (z <=> p_other.z == 0) {
				return w <=> p_other.w;
			}
			return z <=> p_other.z;
		}
		return y <=> p_other.y;
	}
	return x <=> p_other.x;
}
```
</details>

Another thing somewhat hurting readability is that this new operator is frequently used with a return type of `auto`, which we don't allow. I'm not advocating for its inclusion (the extra verbosity is nice for custom implementations), just noting that it's a contributing factor. It *is* dumb that these new types aren't just part of the language, but seeing as stuff like `int32_t` technically isn't built in either, wacky types are something C++ just has to live with.

Repiteo commented on platform/android/SCsub:
```diff
@@ -83,7 +83,8 @@ if lib_arch_dir != "":
         str(env["ANDROID_NDK_ROOT"]) + "/sources/cxx-stl/llvm-libc++/libs/" + lib_arch_dir + "/libc++_shared.so"
     )
-    env_android.Command(out_dir + "/libc++_shared.so", stl_lib_path, Copy("$TARGET", "$SOURCE"))
+    # FIXME: `sources/llvm-libc++` removed in NDK 26 and later. Commenting out just so the C++20 builds complete.
+    # env_android.Command(out_dir + "/libc++_shared.so", stl_lib_path, Copy("$TARGET", "$SOURCE"))
```
Highlighting this because I'm gonna need someone from the @godotengine/android team to clarify matters. C++20 support is added to Android starting with NDK26, which I'm now targeting for this PR, but this shared object is no longer here starting with this version. I *believe* this is deliberate, as the NDK26 changelog[^1] says this:
> libc++ has been updated. The NDK's libc++ now comes directly from our LLVM toolchain, so every future LLVM update is also a libc++ update. 

I have no idea what the full ramifications of this are, beyond it presumably being the reason `sources/cxx-stl/llvm-libc++` and `sources/cxx-stl/llvm-libc++abi` were removed. There's still shared libraries available in NDK26 in the prebuilt section — `toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/<arch_alias>-linux-android` — but that section was already available in NDK23 so maybe it wasn't being used for a reason?

[^1]:https://github.com/android/ndk/wiki/Changelog-r26

AThousandShips commented:
For review purposes I think it'd be greatly appreciated to have each feature used be a separate commit

AThousandShips commented on core/math/vector2.h:
```diff
@@ -243,12 +240,11 @@ _FORCE_INLINE_ Vector2 Vector2::operator-() const {
-_FORCE_INLINE_ bool Vector2::operator!=(const Vector2 &p_vec2) const {
-	return x != p_vec2.x || y != p_vec2.y;
+std::partial_ordering Vector2::operator<=>(const Vector2 &p_other) const {
+	if (x <=> p_other.x == 0) {
```
```suggestion
	if ((x <=> p_other.x) == 0) {
```
Generally I think precedence should be just left without parentheses but this operator is still pretty niche so I think this helps readability here, for all

Repiteo replied:
Come to think of it, this use of the three-way operator was redundant. I can just change it to the functionally-equivalent:
```suggestion
	if (x == p_other.x) {
```

AThousandShips replied:
Absolutely less confusing!

Repiteo commented:
> For review purposes I think it'd be greatly appreciated to have each feature used be a separate commit

Shoot, that's a **really** good idea. My brain autopilots to squashing, but something of this scope deserves to be more spread out. Splitting the changes *now* will take some time, but better late than never

AThousandShips commented:
Many of the files are only one kind, for the rest `git add -i` helps a lot when I do such

Repiteo committed:
Core: Bump C++ Standard to C++20

Repiteo commented:
Rebased and reordered! Now there's 5 commits in total, integrating the following – in order:
- Baseline C++20 support (bare minimum needed to compile).
- `likely` & `unlikely` migrated from macros to attributes.
- `Hashable` concept added, added into default hasher.
- Default comparison operators.
- Simple three-way comparison operators.

hpvb commented:
I want to throw my support behind the concept of moving to c++20, in particular it will give us access to: 

https://en.cppreference.com/w/cpp/utility/source_location

Which will allow us to instrument our builds better, allowing us to log *where* our error macros happen. So instead of getting a message that p_index is out of range in cowdata, we could potentially log where the actual call that went out of index instead.

It'll take a bit of doing, but I think that that alone is worth the upgrade.

Repiteo commented:
Oh wow, that looks *extremely* useful! The header looks pretty lightweight, similar to `<initializer_list>`, <sub>(aka: something that should've just been a language feature)</sub> so I think making an exception and adding that header this PR would be entirely reasonable. It'd all depend on how easily it could be integrated of course, but I can see it as part of a broader commit which revamps the error macros with other new C++20 functionality (eg: `<compare>` library functions for mixed-sign indices, concepts for passing classes with `is_null()` functions to `ERR_FAIL_NULL_*` instead of `ERR_FAIL_COND_*(val.is_null())`, etc)

hpvb commented:
> Oh wow, that looks _extremely_ useful! The header looks pretty lightweight, similar to `<initializer_list>`, (aka: something that should've just been a language feature) so I think making an exception and adding that header this PR would be entirely reasonable. It'd all depend on how easily it could be integrated of course, but I can see it as part of a broader commit which revamps the error macros with other new C++20 functionality (eg: `<compare>` library functions for mixed-sign indices, concepts for passing classes with `is_null()` functions to `ERR_FAIL_NULL_*` instead of `ERR_FAIL_COND_*(val.is_null())`, etc)

Maybe we should aim low for this initially then. Do your first 2 commits as separate PRs, at that point we technically have a C++20 codebase, we can then start using `std::source_location` and we can look at the other new features of C++20 on case-by-case basis?

I think there's things in this still that are more controversial than just the switch to a newer standard.

Repiteo commented:
After some local testing, it'd *definitely* be more involved than this PR should reasonably change. While using it as a drop-in replacement for `FUNCTION_STR`/`__FILE__`/`__LINE__` *technically* works, it doesn't appear to solve the issue of error macros reporting their own scope. It'd certainly make for *prettier* output, and we'd be able to pass column info too, but otherwise outer-scope awareness appears to require a **major** rework.

Repiteo commented:
> Maybe we should aim low for this initially then.

That's what I'm already doing tbh. Everything else I've tried to dig into has felt like it goes beyond the scope of drop-in QOL bumps, so the changes remained conservative (albeit wide-reaching). I could *maybe* see myself reverting the three-way comparison commit if people are THAT skeptical (even if its inclusion is an inevitability), but none of the others strike me as intrusive or contentious.

hpvb commented:
> After some local testing, it'd _definitely_ be more involved than this PR should reasonably change. While using it as a drop-in replacement for `FUNCTION_STR`/`__FILE__`/`__LINE__` _technically_ works, it doesn't appear to solve the issue of error macros reporting their own scope. It'd certainly make for _prettier_ output, and we'd be able to pass column info too, but otherwise outer-scope awareness appears to require a **major** rework.

Oh yeah, it is. I did say "It'll take some doing", but because it is a type it can be put in template parameters, so we can schlep it around to the macro site. I wasn't suggesting we do it in this PR, I just think having C++20 is a prerequisite to starting doing it!

RadiantUwU commented:
As im currently working on refactoring Variant, i feel like concepts from C++20 would be extremely helpful allowing us to more easily register them and remove clutter. I'm imagining a system that would only require you to modify the class itself and everything else would handle itself.

Repiteo commented:
As this PR is no longer burdened by adding new integrations, it's functionally ready for review. Removing draft status!

Repiteo commented on thirdparty/thorvg/patches/cpp20-fixes.patch:
```diff
@@ -0,0 +1,24 @@
+diff --git a/thirdparty/thorvg/src/loaders/svg/tvgSvgSceneBuilder.cpp b/thirdparty/thorvg/src/loaders/svg/tvgSvgSceneBuilder.cpp
```
Note that the changes in this patch are now upstreamed in thorvg[^1], so this would be removed when migrating to one of their upcoming releases (likely 1.0.0)

[^1]: thorvg/thorvg#3316

KoBeWi commented:
Some related issues/PRs:
#54058
#75310

mpsalisbury commented:
Adding to the list of c++20 features that are useful, I find [designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers) very handy.

e.g. in a test that generates a struct, the test instances look like this:
	CHECK_PARSE_OK("+10", (FormatSpecPart{ .width = 10, .sign_type = FormatSpec::SIGN_ALWAYS }));
	CHECK_PARSE_OK("<+10", (FormatSpecPart{ .width = 10, .align_type = FormatSpec::ALIGN_LEFT, .sign_type = FormatSpec::SIGN_ALWAYS }));
	CHECK_PARSE_OK("#10", (FormatSpecPart{ .width = 10, .alternate_form = true }));
	CHECK_PARSE_OK("+#10", (FormatSpecPart{ .width = 10, .sign_type = FormatSpec::SIGN_ALWAYS, .alternate_form = true }));
	CHECK_PARSE_OK("<+#10", (FormatSpecPart{ .width = 10, .align_type = FormatSpec::ALIGN_LEFT, .sign_type = FormatSpec::SIGN_ALWAYS, .alternate_form = true }));

There are a dozen fields in that struct, so being able to use named fields to explicitly populate just some of them is very handy.

Ivorforce commented:
Since we're apparently using this thread to add to the c++20 wish list:
- `no_unique_address` could help by allowing us to easily add `Vector`-like `.write` semantics to more types.
- Differentiate C literals from utf8 literals (`char8_t`)
- Fast paths for constants (`consteval`)
- `std::bit_cast` for `constexpr` purposes

Probably more that I forgot :)

hpvb commented:
From the core meeting: We're requesting reviews from the people assigned. If we approve we will merge this for 4.5.

Main questions are around the effects on the C# bindings and GDExtension compatibility.

Ivorforce reviewed (dismissed):
Looks great! I can't find anything to be sceptical about - you've really boiled this PR down to its essentials (except the equality changes, I guess. But that's mostly code removal.)

One thing caught my eye - I'm seeing you adding a few equality comparisons that weren't there previously. Is this because automatic conversions are no longer applied when you're removing the `operator!=` versions?

Repiteo commented on core/object/ref_counted.h:
```diff
@@ -107,8 +104,9 @@ class Ref {
-		return reference != p_r.reference;
+	template <typename T_Other>
+	_FORCE_INLINE_ bool operator==(const Ref<T_Other> &p_r) const {
+		return reference == p_r.ptr();
```
That's something `Ref` as a whole would benefit from, but I can integrate that here for now.

Repiteo commented on modules/mono/mono_gc_handle.h:
```diff
@@ -46,9 +46,12 @@ struct GCHandleIntPtr {
-	_FORCE_INLINE_ bool operator!=(const GCHandleIntPtr &p_other) { return value != p_other.value; }
 
+	// FIXME: C++20 doesn't allow aggregate initializers to bypass deleted constructors.
+	// http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1008r1.pdf
```
It will not cause problems, no. This was relying on a quirk of pre-c++20 construction that bypassed a deleted constructor with `{}` initialization. In practice, this class is only used internally, and making it truly closed off could still be done if desired via friend constructors or a static method, but that felt outside the scope of this PR.

Repiteo commented on core/variant/variant.h:
```diff
@@ -814,8 +814,71 @@ class Variant {
-	bool operator!=(const Variant &p_variant) const;
 	bool operator<(const Variant &p_variant) const;
+
+	// FIXME: Starting with C++20, the above equality operator is no longer sufficient for comparison
```
The short version is implicit conversion operators have finally bitten us in the ass

The long version is that C++20's way of adding new overloads for comparison operators means that there's now a lot more room for ambiguous comparisons. In particular, the synthesized comparisons mean that multiple ways of handling an implicit conversion can occur. This is surprisingly difficult to articulate, but a somewhat straightforward example reveals itself when trying to compile without these operators:
```
platform\windows\display_server_windows.cpp(2970): error C2666: 'Variant::operator ==': overloaded functions have similar conversions
D:\Godot\godot\core/variant/variant.h(816): note: could be 'bool Variant::operator ==(const Variant &) const'
D:\Godot\godot\core/object/ref_counted.h(104): note: or 'bool Ref<Resource>::operator ==(const Ref<Resource> &) const' [synthesized expression 'y == x']
platform\windows\display_server_windows.cpp(2970): note: while trying to match the argument list '(const T, const Ref<Resource>)'
        with
        [
            T=Variant
        ]
platform\windows\display_server_windows.cpp(2970): error C2666: 'Variant::operator ==': overloaded functions have similar conversions
D:\Godot\godot\core/variant/variant.h(816): note: could be 'bool Variant::operator ==(const Variant &) const'
D:\Godot\godot\core/math/vector2.h(160): note: or 'bool Vector2::operator ==(const Vector2i &) const' [synthesized expression 'y == x']
D:\Godot\godot\core/math/vector2.h(247): note: or 'bool Vector2::operator ==(const Vector2 &) const' [synthesized expression 'y == x']
platform\windows\display_server_windows.cpp(2970): note: while trying to match the argument list '(const T, const Vector2)'
        with
        [
            T=Variant
        ]
```
The relevant code in question:
```cpp
void DisplayServerWindows::cursor_set_custom_image(const Ref<Resource> &p_cursor, CursorShape p_shape, const Vector2 &p_hotspot) {
	_THREAD_SAFE_METHOD_

	ERR_FAIL_INDEX(p_shape, CURSOR_MAX);

	if (p_cursor.is_valid()) {
		RBMap<CursorShape, Vector<Variant>>::Element *cursor_c = cursors_cache.find(p_shape);

		if (cursor_c) {
			if (cursor_c->get()[0] == p_cursor && cursor_c->get()[1] == p_hotspot) {
...
```
In all likelyhood, we've probably been dodging several ambiguous bullets on live and/or are already footgunning ourselves without even realizing it. This entire setup will need a more thorough review, likely accompanied with a re-evaluation of the implicit conversion of `Variant`, but this is very much a stopgap solution to maintain the same comparison functionality as llive.

Repiteo commented:
> One thing caught my eye - I'm seeing you adding a few equality comparisons that weren't there previously. Is this because automatic conversions are no longer applied when you're removing the `operator!=` versions?

It's swapping out a "secondary" comparison operator for a "primary" one. C++20 makes synthesized operators from the primary ones, so it achieves the same result as before but allows for equality operation as well. A lot of this is admittedly more bloated than it should be, as many of the changes were made with defaulted comparison operators & three-way comparison `<=>` in mind, but that'll be for a followup PR which explicitly adds those new features

Repiteo commented:
Revisited the structure after the core meeting, and managed to *significantly* decrease the number of changed files/lines. Now there's no more removed operators; only additional operators to account for the ambiguous cases.
- Before: 79 files, 568 lines
- After: 22 files, 317 lines

I don't believe that there's any real way to circumvent the Variant operator bloat, as it otherwise requires a VERY invasive number of changes. As the comment notes, it's an area that can be revisited with a more general pass later.

dsnopek reviewed (approved):
This looks good to me! There's nothing that seems particularly dangerous or problematic for the areas of the engine I usually work in

Ivorforce reviewed (approved):
I appreciate that you got it even more minimal!

raulsntos reviewed (commented):
I tried compiling this branch and I'm getting these warnings:

```
modules/mono/csharp_script.cpp:1110:59: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'GCHandleIntPtr' and 'GCHandleIntPtr') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
        if (!r_gchandle.is_released() && r_gchandle.get_intptr() == p_gchandle_to_free) { // Do not lock unnecessarily
                                         ~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~
modules/mono/mono_gd/../mono_gc_handle.h:48:22: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
        _FORCE_INLINE_ bool operator==(const GCHandleIntPtr &p_other) { return value == p_other.value; }
                            ^
modules/mono/csharp_script.cpp:1112:60: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'GCHandleIntPtr' and 'GCHandleIntPtr') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
                if (!r_gchandle.is_released() && r_gchandle.get_intptr() == p_gchandle_to_free) {
                                                 ~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~
modules/mono/mono_gd/../mono_gc_handle.h:48:22: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
        _FORCE_INLINE_ bool operator==(const GCHandleIntPtr &p_other) { return value == p_other.value; }
                            ^
modules/mono/csharp_script.cpp:1120:55: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'GCHandleIntPtr' and 'GCHandleIntPtr') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
        if (!gchandle.is_released() && gchandle.get_intptr() == p_gchandle_to_free) { // Do not lock unnecessarily
                                       ~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~
modules/mono/mono_gd/../mono_gc_handle.h:48:22: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
        _FORCE_INLINE_ bool operator==(const GCHandleIntPtr &p_other) { return value == p_other.value; }
                            ^
modules/mono/csharp_script.cpp:1122:56: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'GCHandleIntPtr' and 'GCHandleIntPtr') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
                if (!gchandle.is_released() && gchandle.get_intptr() == p_gchandle_to_free) {
                                               ~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~
modules/mono/mono_gd/../mono_gc_handle.h:48:22: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
        _FORCE_INLINE_ bool operator==(const GCHandleIntPtr &p_other) { return value == p_other.value; }
                            ^
```

However, it finishes building and everything appears to work as expected.

raulsntos reviewed (approved):
Not getting any warnings now, and I see no problems with this for the .NET area.

Calinou reviewed (commented):
Tested locally on Windows 11 24H2 with MSVC 2022, it works as expected. I've also tested an Ubuntu 22.04 Docker container and Godot can still be built.

Note that this makes it no longer possible to compile Godot on Ubuntu 20.04 (unless you install a more recent GCC/Clang through third-party repositories), but 20.04 is EOL since April 2025 anyway.

Code looks good to me.

Some notes:

- Is MSVC 2019 tested (on its latest update) to make sure it works? I don't have it installed, but I don't remember if it has full C++20 support.
- Using Apple's Clang, is it still possible to compile on old macOS versions like 10.15 or 11? We still support 10.15 currently, although only with the Compatibility renderer. It's not a big deal if it can't be compiled on 10.15 anymore, as long as binaries compiled on newer macOS versions still work there.

Benjamin-Dobell commented:
I've been using this branch (its various incarnations) almost exclusively for around 6 months. Mostly targeting macOS and Android.

Re: macOS, I haven't changed anything from the defaults, the resulting binaries are currently minimum macOS 11.0:

```
Load command 10
      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 1
    minos 11.0
      sdk 15.5
   ntools 1
     tool 3
  version 1167.5
  ```

macOS 11 was released _just_ over 5 years ago and can be installed on devices that are over 10 years old (https://support.apple.com/en-au/103111). Obviously this isn't my call to make, and I don't know what the current minimum version is, but requiring macOS 11 seems reasonable for a runtime constraint.

Not exactly sure about the minimum macOS version required to build. I believe `macos-latest` (Github Actions) was pointing at macOS 14 (now two years old) when this PR was initially opened. So we've at least got a couple of years of support for building. I'm still running macOS 15 (one year old) myself.

CoraBlack commented:
That is great! Keep up with the latest C++ trends!

bruvzg reviewed (approved):
Tested various builds, and did not encounter any issues. Code looks good to me. Is there anything that still blocking this?
